spotbugs/src/main/java/edu/umd/cs/findbugs/detect/FindHidingSubClass.java

```diff
@@ -0,0 +1,142 @@
+        for (JavaClass superClass : superClasses) {
+            Method[] superMethods = superClass.getMethods();
+            for (Method method : methods) {
+                // Careful!!! regarding the order of the conditions applied here
+                // Taking advantage of short circuit evaluation here by placing !isMainMethod(method)
+                if (method.isStatic() && !method.isPrivate() && !isMainMethod(method)) {
+                    //I check for the exceptional cases of inner class using three auxiliary private methods.
+                    if (isConstructor(method) || isHidingInnerClass(method) || isAutoGeneratedMethod(method)) {
+                        continue;
+                    }
+                    //I check either the subclass method is hiding the superclass method. If yes I report the bug.
+                    if (Arrays.asList(superMethods).contains(method)) {
+                        bugReporter.reportBug(new BugInstance(this, "HSBC_HIDING_SUB_CLASS", NORMAL_PRIORITY)
+                                .addClass(subClass.getClassName())
+                                .addMethod(subClass, method)
+                                .addClass(superClass.getClassName())
+                                .addSourceLine(SourceLineAnnotation.fromVisitedInstruction(subClass, method, 0)));
+                    }
+                }
+            }
+        }
+    }
+
+    @Override
+    public void report() {
+    }
+
+    /**
+     *This method checks for the inner class exceptional cases.
+     *As whenever there an inner class, '.access$' methods are created hiddenly to access the outer class attributes.
+     */
+    private boolean isHidingInnerClass(Method method) {
+        return method.getName().contains("access$");
+    }
+
+    /**
+     *This method checks for the autoGenerated methods for an inner class - exceptional cases.
+     *As whenever there an inner class, '.class$' methods are created hiddenly.
+     */
+    private boolean isAutoGeneratedMethod(Method method) {
+        return method.getName().contains("class$");
+    }
+
+    /**
+     * This method is here to check the exceptional case of Constructors
+     */
+    private boolean isConstructor(Method method) {
+        return Const.STATIC_INITIALIZER_NAME.equals(method.getName()) || Const.CONSTRUCTOR_NAME.equals(method.getName());
+    }
+
+    /**
+     * This method checks for the exceptional case of main method.
+     * As we know main method always have the signature "public static void main(String[] args)".
+     * It is static but usually public class have its own main method as its entry point.
+     * Therefore, it is not an error caused by Programming but a utility to provide UI to user.
+     */
+    private boolean isMainMethod(Method method) {
+        return method.isPublic() && method.isStatic() && method.getReturnType().equals(BasicType.VOID)
```

Comment:
Things are getting more complicated with [JEP 445](https://openjdk.org/jeps/445), can you please adjust for the new allowed main methods?

Reply:
I have made the condition a bit relaxed to accept `void main(){...}` as the main method. I kept the previous implementation there so that reader can get idea why is it so. If you need me to delete that, let me know.

Reply:
You shouldn't keep the unnecessary parts of the previous implementation, but you could still check the argument types (since not every argument is accepted) and the visibility.

Reply:
I have fixed the condition.